cmake/checks/cpu_fp16.cpp

```diff
@@ -15,12 +15,20 @@ int test()
 #include "arm_neon.h"
 int test()
 {
+#ifdef __ARM_FEATURE_FP16_VECTOR_ARITHMETIC
     const float src[] = { 0.0f, 0.0f, 0.0f, 0.0f };
     short dst[8];
     float32x4_t v_src = *(float32x4_t*)src;
     float16x4_t v_dst = vcvt_f16_f32(v_src);
-    *(float16x4_t*)dst = v_dst;
+    float16x8_t x = vcombine_f16(v_dst, v_dst);
+
+    x = vfmaq_laneq_f16(x, x, x, 7);
+    x = vfmaq_f16(x, x, x);
+
     return (int)dst[0];
+#else
+    return 0;
```

Comment:
Do we really should just pass compilation check in that case here?

Reply:
I'm not very familiar with compiling.
The `vfmaq_laneq_f16` can only be used when `__ARM_FEATURE_FP16_VECTOR_ARITHMETIC` is defined. 
What's your opinion on this part code?

Reply:
It is a compilation check (check of compiler features and compiler flags).

If compiler is old and doesn't provide `__ARM_FEATURE_FP16_VECTOR_ARITHMETIC` but provide "basic" FP16 support then this check should pass. Currently it doesn't try to compile any FP16 code (just `return 0`). Finally it would fail during compilation of OpenCV code if there is no `__ARM_FEATURE_FP16_VECTOR_ARITHMETIC`.

According to `haveFP16_ARM()`, perhaps we need a new `FP16_ARMv82` compiler feature with separate `cmake/checks/cpu_fp16_armv82.cpp` besides of `NEON` and `FP16`.

At least it should include definition check in a reverse form:

```
#ifndef __ARM_FEATURE_FP16_VECTOR_ARITHMETIC
#error "__ARM_FEATURE_FP16_VECTOR_ARITHMETIC must be defined"
#endif
```

Not sure if we need runtime check for that feature at all (`haveFP16_ARM()`).
Anyway runtime checks of CPU features should be performed in system.cpp/HWFeatures (but without direct usage of compiler instructions). Also such detection could be skipped (assumed that baseline is always available) - similar to `CV_FP16`:
```
        #if CV_FP16
        CV_LOG_INFO(NULL, "- FP16 instructions is enabled via build flags");
        have[CV_CPU_FP16] = true;
```
`system.cpp` doesn't support `longjmp()` approach for proper detection (as there is baseline compiler flags only).

Reply:
> perhaps we need a new FP16_ARMv82 compiler feature with separate cmake/checks/cpu_fp16_armv82.cpp besides of NEON and FP16.

I agree with this. But I don't have much time to spend on it recently. I think this patch will have no harm since the `CV_FP16` macro is not used before this patch.
How about leaving this into the future to optimize?

> Not sure if we need runtime check for that feature at all (haveFP16_ARM()).

The runtime check is supplied for opencv binary. Compiler opencv arm binary once, and run every arm chip.

Reply:
@opencv-alalek, FP16 support without SIMD arithmetic support does not make any sense, since C/C++ is not quite able to handle this type properly. I suggest to modify definition of CV_FP16 on ARM as suggested in this PR.